tests/Makefile.sub

```diff
@@ -1,58 +1,63 @@
-
-jlm-valgrind-check: $(JLM_BUILD)/tests/test-runner
-	@rm -rf $(JLM_ROOT)/check.log
-	@FAILED_TESTS="" ; \
-	for TEST in $(TESTS); do \
-		$(TESTLOG) -n "$$TEST: " ; if valgrind --leak-check=full --error-exitcode=1 $(JLM_BUILD)/tests/test-runner $$TEST >>$(JLM_ROOT)/check.log 2>&1 ; then $(TESTLOG) pass ; else $(TESTLOG) FAIL ; FAILED_TESTS="$$UNEXPECTED_FAILED_TESTS $$TEST" ; fi ; \
-	done ; \
-	set -e ; \
-	if [ "x$$FAILED_TESTS" != x ] ; then printf '\033[0;31m%s\033[0m%s\n' "Failed valgrind-tests:" "$$FAILED_TESTS" ; exit 1 ; else printf '\033[0;32m%s\n\033[0m' "All valgrind-tests passed" ; fi ; \
 
-jlm-check-clean:
-	rm -rf $(JLM_BUILD)/tests
-	rm -f $(JLM_ROOT)/ctests.log
-	rm -f $(JLM_ROOT)/utests.log
-	rm -f $(JLM_ROOT)/check.log
+$(eval $(call common_library,libjlmtest))
+
+################################################################################
+# Rules to compile test cases with jlc and run them
+
+JLC_COMPILE_TESTS = \
+	tests/c-tests/test-array-struct \
+	tests/c-tests/test-assertion \
+	tests/c-tests/test-binary \
+	tests/c-tests/test-call-by-value \
+	tests/c-tests/test-casts \
+	tests/c-tests/test-complex \
+	tests/c-tests/test-conditional \
+	tests/c-tests/test-constant-array \
+	tests/c-tests/test-do-while \
+	tests/c-tests/test-extractvalue \
+	tests/c-tests/test-factorial \
+	tests/c-tests/test-fctarray \
+	tests/c-tests/test-fctptr \
+	tests/c-tests/test-fibonacci \
+	tests/c-tests/test-fp-arithmetic \
+	tests/c-tests/test-fp-comparison \
+	tests/c-tests/test-fprintf \
+	tests/c-tests/test-gblfctarray \
+	tests/c-tests/test-global-variable \
+	tests/c-tests/test-irreducible \
+	tests/c-tests/test-long-double \
+	tests/c-tests/test-loop-return \
+	tests/c-tests/test-main \
+	tests/c-tests/test-memcpy \
+	tests/c-tests/test-printf \
+	tests/c-tests/test-ptr-comparison \
+	tests/c-tests/test-recfct \
+	tests/c-tests/test-struct2 \
+	tests/c-tests/test-struct \
+	tests/c-tests/test-switch \
+	tests/c-tests/test-valist \
+	tests/c-tests/test-zeroinitialization
+
+$(patsubst %, $(BUILD_OUT_PREFIX)%.jlm, $(JLC_COMPILE_TESTS)): $(BUILD_OUT_PREFIX)%.jlm : %.c $(BUILD_OUT_PREFIX)jlc
+	@mkdir -p $(dir $@)
+	@cp $< $$(dirname $@)/$$(basename $<)
+	$(BUILD_OUT_PREFIX)jlc -Wall -Werror -O3 -o $@ $$(dirname $@)/$$(basename $<)
+
+GENERATED_FILES+=$(patsubst %, $(BUILD_OUT_PREFIX)%.jlm, $(JLC_COMPILE_TESTS))
+GENERATED_FILES+=$(patsubst %, $(BUILD_OUT_PREFIX)%.c, $(JLC_COMPILE_TESTS))
+GENERATED_FILES+=$(patsubst %, $(BUILD_OUT_PREFIX)%.jlm, $(JLC_COMPILE_TESTS))
+GENERATED_FILES+=$(patsubst %, $(BUILD_OUT_PREFIX)%.o, $(JLC_COMPILE_TESTS))
+
+$(patsubst %, run-%, $(JLC_COMPILE_TESTS)): run-%: $(BUILD_OUT_PREFIX)%.jlm
+	$<
+
+# Run the jlc-compiled binaries as part of tests. This may
+# be problematic when cross-compiling for foreign targets,
+# but leave this to be addressed another day.
+check: $(patsubst %, run-%, $(JLC_COMPILE_TESTS))
```

Comment:
Is there any way to print a little summary at the end. I see that this runs the C tests and the unit tests, but I had to manually redirect the output into a file and inspect the file to see that "everything" is running. Before, we had a little message saying that all tests passed. It was very convenient.

The other "problem" now is that `make check` fails immediately if one test fails without running the rest of the tests at all. Before, it ran all tests and printed me the ones that failed, and I could choose which one to attack. I do not see this as a huge issue, but it basically gave me a choice from which "angle to attack" a problem. If this is not easily to do, then ignore it, please.

Reply:
You can always run "make -i check". I generally like to also parallelize unit test execution. What is going to be useful later is that unit tests should provide output which assertion failed where, such that vim and various IDEs understand it and can correlate the source line. I modified the output of the unit test running command in the makefile so you will have at least easily searchable logs telling you which unit test fails. NB that "make check" exits with an error code if any test fails, is that insufficient?

Reply:
The failure or success is still not as easily visible as before (I think), but good enough. The modification improves it significantly. It should be sufficient that "make check" exits with an error code.